### DISCUSSION TITLE
Doing HTTP call from custom datasource to GitHub results in 401

### DISCUSSION THREAD

**Discussion by akselleirv** [renovatebot/renovate#24408] [Request Help] [answered]
### How are you running Renovate?

Self-hosted Renovate

### If you're self-hosting Renovate, tell us what version of Renovate you run.

v39.0.5

### If you're self-hosting Renovate, select which platform you are using.

github.com

### Was this something which used to work for you, and then stopped?

Placeholder value, please select the correct response from the dropdown

### Describe the problem

I'm trying to fetch a file from a private repository in a custom datasource:
```
    customDatasources: {
        clusterAddons: {
            defaultRegistryUrlTemplate: "https://api.github.com/repos/<MY_ORG>/<MY_REPO>/contents/charts/addon_versions_index.json",
            transformTemplates: [
                "{\"releases\": [{\"version\": $base64decode(content).addonVersionsIndex.{{ packageName }} }]}"
            ]
        }
    }
```

But it results in a 401. 

In my GitHub action pipeline I'm able to curl the file using the same token as Renovate is using:
```
      - name: Generate token
        id: generate_token
        uses: actions/create-github-app-token@v1
        with:
          app_id: ${{ secrets.APPLICATION_ID }}
          private_key: ${{ secrets.APPLICATION_PRIVATE_KEY }}

      - name: Use API
        env:
          MY_TOKEN: ${{ steps.generate_token.outputs.token }}
        run: |
          curl -L \
            -H "Accept: application/vnd.github+json" \
            -H "Authorization: Bearer $MY_TOKEN" \
            -H "X-GitHub-Api-Version: 2022-11-28" \
            https://api.github.com/repos/<MY_ORG>/<MY_REPO>/contents/charts/addon_versions_index.json
```

And it is able to get pull requests from this same repository the file is located in.

### Relevant debug logs

<details><summary>Logs</summary>

```
       "urls": {
         "https://api.github.com/graphql (POST,200)": 32,
         "https://api.github.com/repos/<MY_ORG>/<MY_REPO>/branches/main/protection (GET,404)": 1,
         "https://api.github.com/repos/<MY_ORG>/<MY_REPO>/contents/charts/addon_versions_index.json (GET,401)": 1,
         "https://api.github.com/repos/<MY_ORG>/<MY_REPO>/pulls (GET,200)": 2,
         "https://api.github.com/repos/grafana/helm-charts (GET,200)": 1,
         "https://api.github.com/repos/grafana/helm-charts/git/trees/main (GET,200)": 1,
         "https://grafana.github.io/helm-charts/index.yaml (GET,200)": 1
       },
```

</details>

### Have you created a minimal reproduction repository?

Placeholder value, please select the correct response from the dropdown

**Comment by rarkins** [comment]
How is this different from #24377 ?

**Reply by akselleirv** [reply]
It returned a 404 while this one returns a 401.

**Reply by rarkins** [reply]
Can you paste the full error message, if it's in the logs? Or post what's available if it's not

**Reply by akselleirv** [reply]
[logs.txt](https://github.com/renovatebot/renovate/files/12597583/logs.txt)

The logs, but I'm not sure if it will give you much more.

**Reply by rarkins** [reply]
Looks like we're not including any further details in the logs in this case. But please answer my question below. There's not much point troubleshooting this if it won't work anyway

**Comment by rarkins** [comment, marked as answer]
FWIW I'm not sure if this URL will return what you think it should return. For example look at the results of https://api.github.com/repos/renovatebot/renovate/contents/data/node-js-schedule.json and compare that to https://raw.githubusercontent.com/renovatebot/renovate/main/data/node-js-schedule.json

**Reply by akselleirv** [reply]
I believe it would also work as was handling the response from GitHub in `transformTemplates`:
```
 transformTemplates: [
      "{\"releases\": [{\"version\": $base64decode(content).addonVersionsIndex.{{ packageName }} }]}"
  ]
```

But I changed the URL to https://raw.githubusercontent.com and added the following hostRule:
```
    hostRules: [
        {
            matchHost: "https://raw.githubusercontent.com/<MY_ORG>/<MY_REPO>",
            token: process.env.RENOVATE_TOKEN,
        }
    ]
```


Which worked for me. Thanks again!
